src/marracheck.ml

```diff
@@ -431,89 +425,91 @@ let recover_cover_state ~(selection : PkgSet.t) (cover_state : Cover_state.t) =
-      (* This initializes a fresh cover_state repository; it contains
-         no data at this point (i.e. [cover_state_repo.head = None]) *)
-      let cover_state_repo =
-        Repo.load_and_clean
-          ~path:switch_state.cover_state_repo.path
-          ~load:(fun ~dir:_ -> assert false (* there is no data to load *)) in
-      (* Add some data to the directory (the initial cover state for our
-         packages selection). *)
-      let cover_state = Cover_state.create
-          ~dir:switch_state.cover_state_repo.path
-          ~timestamp:repo_timestamp in
-      Repo.commit_new_head ~sync:Cover_state.sync "Initial cover state"
-        { cover_state_repo with head = Some cover_state }
-  in
-  { switch_state with cover_state_repo }
-
+  if State.(exists st @@ cover_state_path ~switch)
+  && State.(read st @@ timestamp_path ~switch) = repo_timestamp
+  then begin
+    log "Found an existing cover state for the current timestamp";
+    let has_changed =
+      recover_cover_state ~selection ~st ~switch in
+    if not has_changed then
+      (* Avoid polluting the git history with identity commits *)
+      ()
+    else
+      State.(commit st (cover_state_path ~switch) ~msg:"Update cover state")
+  end else begin
+    (* Start over with a fresh cover state for the current timestamp *)
+    if State.(exists st @@ cover_state_path ~switch) then begin
+      log "Existing cover state is for an old repo timestamp";
+      (* There is an existing cover_state, but its timestamp does not match
+         the one of the repository. Retire the cover_state
+         before creating a new one *)
+      retire_cover_state ~st ~switch;
+    end;
+    log "Initialize a fresh cover state";
+    (* This initializes a fresh cover_state repository; it contains
+       no data at this point (i.e. [cover_state_repo.head = None]) *)
+    State.(mkdir st @@ cover_state_path ~switch);
+    (* Initialize the directory. *)
+    (* NOTE: we need to provide the timestamp explicitly, but for the rest we
+       could have an [autoinit] function that would initialize the rest using
+       the default values from the schema.
+
+       Alternatively, we could make it so [State.read] auto-creates files
+       using default values when trying to read from a file that does not
+       exist (currently this auto-creation is only done once by [State.load]).
+       However this becomes a bit tricky if one needs to do it in depth and
+       create directories...
+
+       For now it seems simpler to require that when one creates a directory,
+       one manually initializes it so that it satisfies the schema. *)
+    State.(write st @@ timestamp_path ~switch) repo_timestamp;
+    State.(write st @@ past_elts_path ~switch) [];
+    State.(write st @@ cur_plan_path ~switch) None;
+    State.(write st @@ cur_report_path ~switch) [];
+    State.(write st @@ uninst_path ~switch) PkgSet.empty;
+    State.(commit st ~msg:"Initial cover state" @@ cover_state_path ~switch)
```

Comment:
Proposal: first write `timestamp_path`, and then call `State.(prepare st cover_state_path ~switch)`.